src/MinimumNodeService.cpp

```diff
@@ -383,104 +384,43 @@ Processed MinimumNodeService::handleMessage(unsigned int opc, CANFrame *msg)
-          }          
-          return PROCESSED;
-          
-        case MODE_SETUP:
-          if (nn != 0)   // Not for this node
-          {
-            return PROCESSED;
-          } 
-          else
-          {
-            controller->sendGRSP(OPC_MODE, getServiceID(), CMDERR_INV_CMD);            
-          }
-          return PROCESSED;
-          
-        case MODE_NORMAL:
-          switch (newMode)
-          {
-          case MODE_SETUP:
+          case 0x00:
+          // Request Setup
             controller->sendGRSP(OPC_MODE, getServiceID(), GRSP_OK);
             initSetupFromNormal();
-            break;
-          
-          case MODE_LEARN:
-            //DEBUG_SERIAL << F("> MODE -- Learn") << endl;
-            controller->setLearnMode(newMode);
-            instantMode = newMode;
-            controller->sendGRSP(OPC_MODE, getServiceID(), GRSP_OK);
-            break;
-          
-          case MODE_NHEARTB:          
-            //DEBUG_SERIAL << F("> MODE -- request no Heartbeats") << endl;
-            noHeartbeat = true;
-            instantMode = newMode;
-            break;
-          }
-          return PROCESSED;
-          
-        case MODE_LEARN:
-          if (newMode == MODE_SETUP)
-          {
-            controller->sendGRSP(OPC_MODE, getServiceID(), CMDERR_INV_CMD); 
-          }
-          if (newMode == MODE_NORMAL)
-          {
-            controller->setLearnMode(newMode);
-            if (noHeartbeat == true)
-            {
-              instantMode = MODE_NHEARTB;
-            }
-            else
-            {
-              instantMode = newMode;
-            }
-          } 
-          return PROCESSED;
-          
-        case MODE_NHEARTB:
-          if (newMode == MODE_SETUP)
-          {
-            noHeartbeat = false;
-            controller->sendGRSP(OPC_MODE, getServiceID(), GRSP_OK);
-            initSetupFromNormal();
-          }
-          if (newMode == MODE_NORMAL)
-          {
+            return PROCESSED;
+            
+          case 0x0C:
+          // Turn on Heartbeat
             noHeartbeat = false;
-            instantMode = newMode;
-            controller->sendGRSP(OPC_MODE, getServiceID(), GRSP_OK);
+            module_config->setHeartbeat(!noHeartbeat);
+            return PROCESSED;
+            
+          case 0x0D:
+          // Turn off Heartbeat
+            noHeartbeat = true;
+            module_config->setHeartbeat(!noHeartbeat);
+            return PROCESSED;
           }
-          if (newMode == MODE_LEARN)
-          {
-            controller->setLearnMode(newMode);
-            instantMode = newMode;
-            controller->sendGRSP(OPC_MODE, getServiceID(), GRSP_OK);
-          } 
+          // if in MODE_NORMAL but none of the above commands, let another service see message
+          break;  
+        }
+        else
+        {
           return PROCESSED;
         }
       }
-      
+      else
+      {
       return PROCESSED;
```

Comment:
Thinking about these return statements. I don't think we should stop processing here. Some other service might want to do something for this MODE code.

Reply:
This is a default for when we are not MODE_NORMAL.  If we are not Normal than no other services should be doing anything.
Thinking overnight, I think that the "break;" at line 413 should be a default case and not just be hanging.

Reply:
OK, I read what you are saying. It does make sense that MODE is only used in Normal mode. 
I am just thinking that in the future things might change. But we'll review the code in that case.

Regarding the "break", I think you can replace it with "return NOT_PROCESSED;". The break relies on such a return statement at the end of the function. It is clearer to have the return NOT_PROCESSED here.

Reply:
I would prefer "return NOT_PROCESSED" as it is much clearer as to what is happening.